Sub Node alternative

fmrico opened the pull request:
Hi, 

This PR is related to issue #531 . There is a previous PR (#578), and this PR is an alternative to it.

Best

fmrico commented:
Detailed description of the PR:

- This PR allows you to have the concept of NodeHandle discussed in #531 
- It provides a mechanism (Sub Node) to refer to the same Node, but having an extra namespace. From the comments in #578 y [here](https://github.com/ros2/rclcpp/issues/531#issuecomment-412235400), it is preferred to extend the Node functionality instead of creating a new SubNode, NodeProxy or NodeHandler class. Duplicated code is avoided.
- You can get this functionality by method `create_sub_node(original_node, extended_ns)`, getting another way to access the `original_node` (not creating a new one) with a extended_ns.
- get_namespace() returns `node_base_` namespace, so it is not internally extended.
- It checks if the extended namespace is valid.
- Absolute topic names ignore extended namespace (debatable, as everything)
- Example: 

``` 
  node_1_ptr = rclcpp::Node::make_shared("node_1");
  std::cerr<<"Namespace = "<<node_1_ptr->get_namespace()<<std::endl;
  auto sub_1_1 = node_1_ptr->create_subscription<std_msgs::msg::String>("/n1_s1",topic_callback);
  auto sub_1_2 = node_1_ptr->create_subscription<std_msgs::msg::String>("n1_s2",topic_callback);
  auto sub_1_3 = node_1_ptr->create_subscription<std_msgs::msg::String>("~/n1_s4",topic_callback);

  auto node_s1_ptr = node_1_ptr->create_sub_node("sub_ns_1");
  std::cerr<<"Namespace = "<<node_s1_ptr->get_namespace()<<std::endl;
  auto sub_s1_1 = node_s1_ptr->create_subscription<std_msgs::msg::String>("/sn1_s1",topic_callback);
  auto sub_s1_2 = node_s1_ptr->create_subscription<std_msgs::msg::String>("sn1_s2",topic_callback);
  //auto sub_s1_3 = node_s1_ptr->create_subscription<std_msgs::msg::String>("~/sn1_s4",topic_callback);  // Invalid 

  auto node_ss1_ptr = node_s1_ptr->create_sub_node("subsub_ns_1");
  std::cerr<<"Namespace = "<<node_ss1_ptr->get_namespace()<<std::endl;
  auto sub_ss1_1 = node_ss1_ptr->create_subscription<std_msgs::msg::String>("/ssn1_s1",topic_callback);
  auto sub_ss1_2 = node_ss1_ptr->create_subscription<std_msgs::msg::String>("ssn1_s2",topic_callback);
  //auto sub_s1_3 = node_s1_ptr->create_subscription<std_msgs::msg::String>("~/sn1_s4",topic_callback);  // Invalid 
```

This creates next topics:

```
/n1_s1
/n1_s2
/node_1/n1_s4
/parameter_events
/sub_ns_1/sn1_s1
/sub_ns_1/sn1_s2
/sub_ns_1/subsub_ns_1/ssn1_s1
/sub_ns_1/subsub_ns_1/ssn1_s2
```

best

wjwwood commented:
Thanks for iterating on it, I'll have a look as soon as I can, or maybe someone else can have a look sooner than me. Thanks for your patience.

wjwwood commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -257,7 +273,9 @@ Node::get_parameter_or_set(
-  bool got_parameter = get_parameter(name, value);
+  std::string extended_name = get_extended_name(name);
+
+  bool got_parameter = get_parameter(extended_name, value);
```
I'm not sure about this for parameter names... I think they use the `.` separator at the moment if I'm not mistaken (https://github.com/ros2/rclpy/blob/5b95169e6f4bc05e3052e22ccbe036df69898f8a/rclpy/rclpy/parameter.py#L20), though I think that's been a point of discussion in the past.

@tfoote @dirk-thomas can I get a consult on this from one or both of you guys? Specifically "should a concept like `NodeHandle` from ROS 1 affect the name when getting or setting a parameter"?

fmrico commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -84,6 +85,52 @@ Node::Node(
+  if (other.extended_namespace_ == "") {
+    extended_namespace_ = extended_namespace;
+  } else {
+    extended_namespace_ = other.extended_namespace_ + "/" + extended_namespace;
```
Hi @wjwwood ,

In this case it throws a rclcpp::exceptions::InvalidNamespaceError. Is it correct, or should I fix the case of "//" in namespace?

fmrico replied:
This is taken into account in the last commits. I have assured that `extended_namespace` (now `sub_namespace`) never starts or ends by `/`, and I check concatenations between `namespace_` and `sub_namespace`

When namespace ends in `/`, it continues throwing `rclcpp::exceptions::InvalidNamespaceError`

fmrico commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -84,6 +85,52 @@ Node::Node(
+    if (rmw_ret == RMW_RET_INVALID_ARGUMENT) {
+      throw_from_rcl_error(RCL_RET_INVALID_ARGUMENT, "failed to validate subnode namespace");
+    }
+    throw_from_rcl_error(RCL_RET_ERROR, "failed to validate subnode namespace");
```
ok. I will use it as soon as it is available

fmrico replied:
I could use it as soon it became public

wjwwood commented:
@fmrico I was trying to fixup your commits (it looked like commits from master were on this branch) but I think I somehow also eliminated your comment about the namespace exception on accident.

Can you repost it? Or was it already addressed? I see some tests related to `//` already in 65330d3.

jacobperron commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -447,9 +452,16 @@ class Node : public std::enable_shared_from_this<Node>
 
+  RCLCPP_PUBLIC
+  Node::SharedPtr
+  create_sub_node(const std::string & extended_namespace_);
```
"subnode" is nice and succinct. And we can refer to the namespace as the "sub namespace", rather than "extended". 

I'm also okay with another option. In any case, I agree that the naming should be consistent.

fmrico replied:
I changed all the `extended_namespace` to `sub_namespace`.

On how to call the object returned by this function, I am receptive to any option that seems appropriate.

fmrico commented:
Hi @wjwwood ,

I deleted the messages because yesterday I could not upload all the changes, and it was not consistent with the comments. Maybe I should not have deleted them.

I have made changes to take into account the cases that you told me, and I have added many more tests that discovered some errors that I solved.

The only thing I could not solve is to include the sub namespace in the loggers. By now, it uses the real namespace and ignores the sub-namespace.

As @jacobperron suggested, I changed extended_namespace to sub_namespace. Actually, it is more appropriate.

fmrico commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -84,6 +85,52 @@ Node::Node(
+    throw rclcpp::exceptions::InvalidNamespaceError(
+            full_namespace.c_str(),
+            rmw_namespace_validation_result_string(validation_result),
+            invalid_index);
```
Fixed!! I always run tests in order to avoid this

fmrico commented on rclcpp/test/test_node.cpp:
```diff
@@ -75,6 +75,75 @@ TEST_F(TestNode, get_name_and_namespace) {
   }
 }
 
+
```
Fixed

fmrico commented on rclcpp/test/test_node.cpp:
```diff
@@ -75,6 +75,75 @@ TEST_F(TestNode, get_name_and_namespace) {
+  }
+}
+
+
```
Fixed

fmrico commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -447,9 +452,16 @@ class Node : public std::enable_shared_from_this<Node>
   RCLCPP_DISABLE_COPY(Node)
 
+  RCLCPP_PUBLIC
+  std::string get_extended_name(const std::string & name) const;
```
I removed this in last commits. This function returns the name of topic/paramenter/service after concatenating the extended namespace

fmrico commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -105,6 +105,11 @@ class Node : public std::enable_shared_from_this<Node>
+  RCLCPP_PUBLIC
+  Node(
+    const Node & other,
+    const std::string & extended_namespace_);
```
Documentation added.

I added more tests that cover many relative/absolute cases.

About `extended_namespace_`, this is a mistake. Actually, in cpp it is correct.

cottsay commented:
It looks like this hasn't seen any movement for a while.

@wjwwood, it looks like @fmrico has addressed some of your feedback from the last round of review. If you could take a second to confirm that those issues are addressed and resolve them, it would be appreciated.

Also re-pinging @sloretz to take a look.

Thanks!

fmrico commented:
Hi @cottsay ,

I pause this contribution some time ago, but I hope to continue this work soon. I think it was near to be finished, but I have to update the code with the changes in other packages since the last commit.

Best

cottsay commented:
Thanks @fmrico - please let us know when this is ready for review again.

fmrico commented:
Hi @cottsay @wjwwood ,

I have just checked that I already resolved the reviews that @wjwwood did. 

This branch has been updated (@wjwwood already did most of the work) to master, and I have just fixed a merge issue. 

It builds with no errors, and I passed all the tests (I already had added tests creation/destruction, namespaces, and subscription/publication, services for sub node).

Please check that everything is correct, or make me the necessary reviews to close this PR.

Best

wjwwood commented:
I was about to push some changes too, oops. I'll have to rebase mine up (I had fixed the merge error too), I should have pushed sooner. I'll try to finish it up tomorrow.

wjwwood commented:
@fmrico I pushed some (somewhat) significant changes, and it would be nice to get your feedback on them. Probably the most controversial thing I did was make it so that `Node::get_namespace()` does not consider the sub-node's sub-namespace, instead returning the Node's original given namespace and adding functions for accessing the sub-namespace and the effective namespace which combines the two.

Also, if anyone on @ros2/team has time, it would be good to get a review on my changes.

wjwwood commented:
Another thing to point out is that Actions will not consider this function, so they should either be updated to consider the sub-namespace (if there is one) or only enforce that when we add a "create_action" method on the Node interface itself (which has been discussed).

One more thing is that this feature would be nice to integrate into the LifecycleNode, but cannot be done easily at the moment. I considered refactoring this more so that it was a "mixin" style thing, which could then just be inherited in both Node and LifecycleNode, but didn't follow through on that.

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6174)](http://ci.ros2.org/job/ci_linux/6174/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2635)](http://ci.ros2.org/job/ci_linux-aarch64/2635/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5089)](http://ci.ros2.org/job/ci_osx/5089/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6030)](http://ci.ros2.org/job/ci_windows/6030/)

clalancette commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -481,6 +489,81 @@ class Node : public std::enable_shared_from_this<Node>
+   * However, if a sub-node (short for subordinate node) is created, then
+   * its sub-namespace will be extended with the sub-namespace given here, and
+   * any entities created with the sub-node object instead of a normal node
+   * object will be affected by the sub-namespace.
```
I'm finding the above paragraph a little difficult to read, partially because it is described in the "negative" sense, partially because it seems to be more like high-level documentation rather than documentation for this particular API, and partially because the terms aren't all defined.  I might suggest something more like the following:

```
Any entities (topics, services, etc.) created with this Node pointer will be extended with the sub-namespace (short for subordinate namespace) given here.  Note that entities created using a "normal" Node pointer will not be affected by this sub-namespace.
```

wjwwood replied:
I can work on the wording, but in your suggestion it's not necessarily a pointer, I think instance is a more appropriate term, and I don't see the reason to avoid the `sub-node` term, as it is described (defined?) in the `create_sub_node()` method.

wjwwood replied:
I rewrote this section in the doc block, please take another look.

I have so much in this particular doc block because I don't know of a better place to describe these concepts right now.

clalancette commented on rclcpp/include/rclcpp/node_impl.hpp:
```diff
@@ -74,11 +86,12 @@ Node::create_publisher(
   return rclcpp::create_publisher<MessageT, Alloc, PublisherT>(
     this->node_topics_.get(),
-    topic_name,
+    extend_name_with_sub_namespace(topic_name, this->get_sub_namespace()),
```
Could we just get rid of `get_sub_namespace()` and get the private variable `this->sub_namespace` directly?  I think that would help with my concerns about the API being confusing (though it would make the tests somewhat less detailed).  Thoughts?

wjwwood replied:
I would not, as I said above, I don't see a reason not to allow access to it, and making the tests harder to implement seems like a pretty significant cost to me.

clalancette commented on rclcpp/src/rclcpp/node.cpp:
```diff
@@ -34,10 +34,59 @@
+  // Assumption is that the existing_sub_namespace does not need checking
+  // because it would be checked already when it was set with this function.
+
+  // check if the new sub-namespace extension is abolsute
```
Nit: `abolsute` -> `absolute`

clalancette commented on rclcpp/test/test_subscription.cpp:
```diff
@@ -107,6 +135,31 @@ TEST_F(TestSubscription, construction_and_destruction) {
+  }
+}
+
 /*
```
Thanks for adding a bunch of tests, that is useful.

The group of tests that I think are missing are the ones for the private node namespace, i.e. the `~` ones from [here](https://github.com/ros2/rclcpp/issues/531#issuecomment-437201989).  There are some above, but none check that the topic/service name properly resolve to the topic name that we expect.

wjwwood replied:
I'll add those tests.

wjwwood commented on rclcpp/include/rclcpp/node.hpp:
```diff
@@ -481,6 +489,81 @@ class Node : public std::enable_shared_from_this<Node>
+   */
+  RCLCPP_PUBLIC
+  const std::string &
+  get_effective_namespace() const;
```
I felt like I had described their difference in the doc blocks, but maybe it needs clarification.

Briefly, `get_namespace()` returns the node's namespace (which is unaffected by any sub-namespaces), `get_sub_namespace()` returns the current sub-namespace, and `get_effective_namespace()` is the namespace that will effectively be used by entities (like pub/sub/client/server) when being created.

I don't expect the user to ever need to call these functions unless they're doing something complicated or their using it to construct some very specific logging functions. `get_effective_namespace()` is purely a convenience, since it's actually not used when creating entities, so I guess we could eliminate it. But the other two, return information that is originally given by the user and that the node needs to store, so I don't see a reason to not allow the user to get the information back out of the node (keep the members private but remove the accessors).

clalancette replied:
> I felt like I had described their difference in the doc blocks, but maybe it needs clarification.
> 
> Briefly, `get_namespace()` returns the node's namespace (which is unaffected by any sub-namespaces), `get_sub_namespace()` returns the current sub-namespace, and `get_effective_namespace()` is the namespace that will effectively be used by entities (like pub/sub/client/server) when being created.
> 
> I don't expect the user to ever need to call these functions unless they're doing something complicated or their using it to construct some very specific logging functions. `get_effective_namespace()` is purely a convenience, since it's actually not used when creating entities, so I guess we could eliminate it. But the other two, return information that is originally given by the user and that the node needs to store, so I don't see a reason to not allow the user to get the information back out of the node (keep the members private but remove the accessors).

Yeah, as mentioned, I basically understood that from reading the code.  My biggest worry is that exposing these APIs to the user is going to result in a lot of answers.ros.org questions of the form: when do I call which of these APIs?  On the other hand, you seem to be saying that these probably won't need to be called except in specialized circumstances, so maybe that concern is unwarranted.  It's not a blocker for me, so I won't press the point further.

wjwwood replied:
I understand the concern, I still would error on the side of exposing information. At some level I think it’s good if this leads to questions about which to call, as that forces people to learn about the options, which I’d prefer to them calling `get_namespace()` but receiving something different from what they assumed. 

Thanks for iterating, you or anyone comes up with a better way to organize it that might be less confusing but still provides access to available data, I’m all ears.

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6193)](http://ci.ros2.org/job/ci_linux/6193/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2648)](http://ci.ros2.org/job/ci_linux-aarch64/2648/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5102)](http://ci.ros2.org/job/ci_osx/5102/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6046)](http://ci.ros2.org/job/ci_windows/6046/)

wjwwood commented:
@clalancette the CI failures are due to something on the farm (I think), so if you could update your review after my last iteration, I'd appreciate it!

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=6205)](http://ci.ros2.org/job/ci_linux/6205/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=2656)](http://ci.ros2.org/job/ci_linux-aarch64/2656/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=5117)](http://ci.ros2.org/job/ci_osx/5117/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=6057)](http://ci.ros2.org/job/ci_windows/6057/)

wjwwood commented:
I’m going to merge this without DCO passing as it was started before the enforcement.

fmrico commented:
Thank for your help, specially @wjwwood , who made a great work improving this PR.

:)

wjwwood commented:
Thanks for try different approaches to find this one, it was a great help :)

addy1997 commented:
#578, #531, and #581 are quite confusing. Which out of these is an alternative for `ros::NodeHandle `?

Actually, I am trying to implement a plugin (relevant code snippet given below)

```
if (!rclcpp::ok()) {

    RCLCPP_FATAL_STREAM("actor", "A ROS node for Gazebo has not been initialized," <<
      "unable to load plugin. Load the Gazebo system plugin " <<
      "'libgazebo_ros_api_plugin.so' in the gazebo_ros package)");
    return;
  }

  this -> dataPtr -> sdf = _sdf;
  this -> dataPtr -> actor = boost::dynamic_pointer_cast < physics::Actor > (_model);
  this -> dataPtr -> world = this -> dataPtr -> actor -> GetWorld();

  this -> dataPtr -> ResetEntity();

  this -> ros_node_ = new ros::NodeHandle();
```

In this code, I am trying to implement `new` node (last line). How should I change it with rclcpp?

wjwwood commented:
@addy1997 This sounds like a question, it would be more appropriate to ask this on answers.ros.org. If you post that link back here, I'll try to answer it.
